### ISSUE TITLE
[red-knot] Wrong possibly-unbound diagnostic for named expression in Boolean chain

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#139]
This seems like a control flow / statically-known-branches bug?
```py
def _(flag: bool):
    if flag and (x := 1):
        x  # possibly-unresolved-reference
```
(https://playknot.ruff.rs/62b1d215-483c-488f-a779-a4856847df52)

**Comment by AlexWaygood** [comment]
Added it to the "GA" milestone as it's important we fix this, but I don't think walruses are common enough to warrant prioritising it. LMK if you disagree!

**Comment by nickdrozd** [comment]
I don't know about others, but I use this pattern all the time. Therefore, currently ty blasts me with false positives.

Here is an example from some code that does algebra stuff:

```python
    def __floordiv__(self, other: Count) -> Count:
        if other == 1:
            return self

        assert isinstance(other, int)

        if (((lgcd := gcd(other, l := self.l)) == 1
                or (rgcd := gcd(other, r := self.r))) == 1
                or (div := gcd(lgcd, rgcd)) == 1):
            return make_div(self, other)

        return ((l // div) + (r // div)) // (other // div)
```

Admittedly this snippet goes a little overboard with assignment expressions. Still, it works fine. ty flags five variables as possibly unresolved, all false positives.

**Comment by gusutabopb** [comment]
I hit this issue when trying out `ty` on an existing code base.


Sample code:

```python
import random

if (random.random() > 0.5) and (m := random.random()):
    # m is always defined, but ty warns it may not be (incorrect)
    print(m)
```
ty output

```
warning: lint:possibly-unresolved-reference: Name `m` used when possibly not defined
 --> r_and_m.py:5:11
  |
3 | if (random.random() > 0.5) and (m := random.random()):
4 |     # m is always defined. ty warns it may not be (incorrect)
5 |     print(m)
  |           ^
  |
info: `lint:possibly-unresolved-reference` is enabled by default
```

Flipping the order of the operators "fixes" the issue though:

```python
import random

if (m := random.random()) and (random.random() > 0.5):
    # m is always defined. ty is silent (correct)
    print(m)
```

ty output

```
All checks passed!
```

**Comment by danielhollas** [comment]
I've tested all the snippets here (both on the playground and latest alpha.8 version) and none of them report any diagnostics. 🎉

**Comment by carljm** [comment]
I think this bug still exists. It was fixed in https://github.com/astral-sh/ruff/pull/18010, but that caused a performance regression and had to be rolled back, and I don't think we've investigated the root cause of the regression yet in order to try to re-land the fix. The reason there are no longer any diagnostics is that we decided to disable the `possibly-unbound-reference` diagnostic by default, since it will inevitably have some false positives.

**Comment by danielhollas** [comment]
>  The reason there are no longer any diagnostics is that we decided to disable the possibly-unbound-reference diagnostic by default, since it will inevitably have some false positives.

Ah, that's what I was missing, thanks!

**Comment by sharkdp** [comment]
Closing this in favor of https://github.com/astral-sh/ty/issues/626, which has a bit more details on `possibly-unresolved-reference`.
